operators/grpc_operators/common/tensor_proto.cpp

```diff
@@ -0,0 +1,167 @@
+    _holoscan_cuda_err;                                                                    \
+  })
+#endif
+
+void TensorProto::gxf_time_to_proto(const nvidia::gxf::Entity& gxf_entity,
+                                    ::holoscan::entity::Timestamp* timestamp) {
+  auto gxf_timestamp = gxf_entity.get<nvidia::gxf::Timestamp>();
+  if (gxf_timestamp) {
+    timestamp->set_acqtime((*gxf_timestamp)->acqtime);
+    timestamp->set_pubtime((*gxf_timestamp)->pubtime);
+  }
+}
+
+void TensorProto::proto_to_gxf_time(nvidia::gxf::Entity& gxf_entity,
+                                    const ::holoscan::entity::Timestamp& timestamp) {
+  auto gxf_timestamp = gxf_entity.add<nvidia::gxf::Timestamp>("timestamp");
+  (*gxf_timestamp)->acqtime = timestamp.acqtime();
+  (*gxf_timestamp)->pubtime = timestamp.pubtime();
+}
+
+void TensorProto::gxf_tensor_to_proto(
+    const nvidia::gxf::Entity& gxf_entity,
+    google::protobuf::Map<std::string, ::holoscan::entity::Tensor>* tensor_map) {
+  auto tensors = gxf_entity.findAll<nvidia::gxf::Tensor, 4>();
+  if (!tensors) { throw std::runtime_error("Tensor not found"); }
+
+  for (auto tensor : tensors.value()) {
+    holoscan::entity::Tensor& tensor_proto = (*tensor_map)[tensor->name()];
+    for (uint32_t i = 0; i < (*tensor)->shape().rank(); i++) {
+      tensor_proto.add_dimensions((*tensor)->shape().dimension(i));
+    }
+    switch ((*tensor)->element_type()) {
+      case nvidia::gxf::PrimitiveType::kUnsigned8:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned8);
+        copy_data_to_proto<uint8_t>(tensor.value(), tensor_proto);
+        break;
+      case nvidia::gxf::PrimitiveType::kUnsigned16:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kUnsigned16);
+        copy_data_to_proto<uint16_t>(tensor.value(), tensor_proto);
+        break;
+      case nvidia::gxf::PrimitiveType::kFloat32:
+        tensor_proto.set_primitive_type(holoscan::entity::Tensor::kFloat32);
+        copy_data_to_proto<float>(tensor.value(), tensor_proto);
+        break;
+      default:
+        throw std::runtime_error("Unsupported primitive type");
+    }
+  }
+}
+
+template <typename T>
+void TensorProto::copy_data_to_proto(const nvidia::gxf::Handle<nvidia::gxf::Tensor>& tensor,
+                                     ::holoscan::entity::Tensor& tensor_proto) {
+  if ((*tensor).storage_type() == nvidia::gxf::MemoryStorageType::kDevice) {
+    void* in_data_ptr = (*tensor).pointer();
+    size_t data_size = (*tensor).bytes_size();
+    std::vector<T> in_data(data_size);
+    CUDA_TRY(cudaMemcpy(in_data.data(), in_data_ptr, data_size, cudaMemcpyDeviceToHost));
+    tensor_proto.set_data(in_data.data(), data_size);
+    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kDevice);
+  } else {
+    tensor_proto.set_data((*tensor).pointer(), (*tensor).size());
+    tensor_proto.set_memory_storage_type(holoscan::entity::Tensor::kSystem);
```

Comment:
Do we need to handle `kHost` here as well? `kHost` is CUDA pinned host memory. Accessing with CPU is slow (uncached reads).

Reply:
Create a separate [issue](https://github.com/nvidia-holoscan/holohub/issues/570) to support `kHost`.